gdx/src/com/badlogic/gdx/utils/Queue.java

```diff
@@ -0,0 +1,219 @@
+ * This collection is not thread-safe. */
+public class Queue<T> {
+
+	/** Contains values waiting in this queue. Pop and push indices go in circle around this array, wrapping at the end. */
+	protected T[] values;
+
+	/** Index to dequeue/remove from. Logically smaller than tail. Unless empty, it points to a valid element inside queue. */
+	protected int head = 0;
+
+	/** Index to enqueue/add to. Logically bigger than head. Usually points to an empty values position, but when full (size ==
+	 * values.length) points to the head. */
+	protected int tail = 0;
+
+	/** Amount of things currently enqueued. */
+	protected int size = 0;
+
+	/** Non resizable queues will have limited capacity.
+	 * @see Queue#add(Object) */
+	public boolean resizable;
+
+	/** Create a new Queue which can hold 15 values without resizing. */
+	public Queue () {
+		this(16, true);
+	}
+
+	/** Create a new Queue which can hold `initialSize` values without resizing. */
+	public Queue (int initialSize, boolean resizable) {
+		// noinspection unchecked
+		this.values = (T[])new Object[initialSize];
+		this.resizable = resizable;
+	}
+
+	/** Create a new Queue which can hold `initialSize` values without resizing. This creates backing array of correct type via
+	 * reflection. Use this only if you are accessing backing array directly.
+	 * <p/>
+	 * NOTE: Worth using only if you know what are you doing. */
+	public Queue (int initialSize, boolean resizable, Class<T> type) {
+		// noinspection unchecked
+		this.values = (T[])ArrayReflection.newInstance(type, initialSize);
+		this.resizable = resizable;
+	}
+
+	/** Enqueue given object. Unless backing array needs resizing, operates in O(1) time.
+	 * @param object can be null
+	 * @return true if added, false if full (can only happen when resizable is false) */
+	public boolean add (T object) {
+		T[] values = this.values;
+
+		if (size == values.length) {
+			if (!resizable) {
+				return false;
+			} else {
+				resize(values.length << 1);// *2
+				values = this.values;
+			}
+		}
+
+		values[tail++] = object;
+		if (tail == values.length) {
+			tail = 0;
+		}
+		size++;
+		return true;
+	}
+
+	/** Optionally resize backing array so adding `additional` amount of entries won't cause resize.
+	 * <p/>
+	 * NOTE: This WILL resize non-resizable arrays. */
+	public void ensureCapacity (int additional) {
+		final int needed = size + additional;
+		if (values.length < needed) {
+			resize(needed);
+		}
+	}
+
+	/** Resize backing array. newSize must be bigger than current size. */
+	private void resize (int newSize) {
+		final T[] values = this.values;
+		final int head = this.head;
+		final int tail = this.tail;
+
+		@SuppressWarnings("unchecked")
+		final T[] newArray = (T[])ArrayReflection.newInstance(values.getClass().getComponentType(), newSize);
+		if (head < tail) {
+			// Continuous
+			System.arraycopy(values, head, newArray, 0, tail - head);
+		} else {
+			// Wrapped
```

Comment:
Just an optimization.
If `head == tail` the buffer can be empty or full.
if it's empty, i.e. `size == 0`, you don't need to copy the whole buffer. So I'd suggest to use `else if (size > 0) {` instead of
`else {`